File: csharp/ConstructBST.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode
{
    class ConstructBST
    {
        public TreeNode BuildTree(int[] preorder, int[] inorder)
        {
            return RecursiveBuildTree(new Queue<int>(preorder), new List<int>(inorder), 0, inorder.Length - 1);
        }

        private TreeNode RecursiveBuildTree(Queue<int> preorder, List<int> inorder, int startIndex, int endIndex)
        {
            if (startIndex == endIndex)
            {
                TreeNode ret = new TreeNode(preorder.Dequeue());
                return ret;
            } else if (startIndex > endIndex)
            {
                return null;
            }
            else
            {
                TreeNode ret = new TreeNode(preorder.Dequeue());
                int index = inorder.IndexOf(ret.val);
                ret.left = RecursiveBuildTree(preorder, inorder, startIndex, index - 1);
                ret.right = RecursiveBuildTree(preorder, inorder, index + 1, endIndex);
                return ret;
            }
        }
    }
}

File: js/ConstructBST.js

function TreeNode(val){
    this.val = val;
    this.left = this.right = null;
}

var buildTree = function(preorder, inorder){
    return recursiveBuildTree(preorder, inorder, 0, inorder.length-1);
};

var recursiveBuildTree = function(preorder, inorder, startIdx, endIdx){
    if(startIdx === endIdx){
        return new TreeNode(preorder.shift());
    } else if (startIdx > endIdx){
        return null;
    } else {
        let ret = new TreeNode(preorder.shift());
        let index = inorder.indexOf(ret.val);
        ret.left = recursiveBuildTree(preorder, inorder, startIdx, index-1);
        ret.right = recursiveBuildTree(preorder, inorder, index+1, endIdx);
        return ret;
    }
};